### ISSUE TITLE
WebGL-renderer: implement preRender and postRender

### ISSUE THREAD

**Issue by MichaelLangbein** [openlayers/openlayers#11463]
### Description

Currently, Heatmaps don't fire  `prerender` or `postrender` events. This is because a.t.m. only `CanvasLayerRenderer` implements the `preRender` and `postRender`  methods. 

Here's a minimum working example: note how only the OSM's render-events are being logged:

```js
const layers = [
    new TileLayer({
        source: new OSM(),
    }),
    new Heatmap({
        source: new VectorSource({
            features: new GeoJSON().readFeatures({
                "type": "FeatureCollection",
                "features": [{
                    "type": "Feature",
                    "properties": {},
                    "geometry": {
                      "type": "Point",
                      "coordinates": [13.38134765625, 52.50953477032727]
                    }
                }]
              })
        })
    })
];

const view = new View({
    center: [15, 52],
    zoom: 7,
    projection: 'EPSG:4326'
});


const map = new Map({
    layers, view, target
});


for (const layer of layers) {
    layer.on('postrender', (event) => {
        console.log(event);
    });
}
```

### Solution

I believe we can literally just copy-paste the source code of `CanvasLayerRenderer`'s `preRender` and `postRender`  methods into `WebGLLayerRenderer`. We'd then only have to call `preRender` and `postRender` from `WebGLPointsLayerRenderer` inside of `renderFrame`. 

Would this be appropriate? It is my understanding that WebGl-support is still somewhat experimental (is that still true?). But if this is something we'd like to see, I'd like to try myself on a pull-request over the weekend. 

Keep up the great work and thanks a lot! :)

**Comment by jahow** [comment]
That sounds absolutely appropriate. Please go ahead with a PR if you can, and don't hesitate to ask here for indications if needed. Thanks!

**Comment by MichaelLangbein** [comment]
Thanks a lot! Working on this now. But I could indeed need a little pointer here.

This is the new `WebGLLayerRenderer.dispatchRenderEvent_` method I've created (You can find the full current state [here](https://github.com/openlayers/openlayers/compare/main...MichaelLangbein:webGlPostRender)): 

```js
  /**
   * @param {import("../../render/EventType.js").default} type Event type.
   * @param {WebGLRenderingContext} context Context.
   * @param {import("../../PluggableMap.js").FrameState} frameState Frame state.
   * @private
   */
  dispatchRenderEvent_(type, context, frameState) {
    const layer = this.getLayer();
    if (layer.hasListener(type)) {
      /**
       * inversePixelTransform: converts from css-pixel-coordinates (rel. to top left) to webgl-vertex-space-coordinates ([-1, 1] x [-1, 1]).
       * [ a c e ]
       * [ b d f ]
       * [ 0 0 1 ]
       */
      const canvas = this.helper.getCanvas();
      const inversePixelTransform = [
        1 / (canvas.width / 2),
        0,
        0,
        -1 / (canvas.height / 2),
        -1,
        1,
      ];

      const event = new RenderEvent(
        type,
        inversePixelTransform,
        frameState,
        context
      );
      layer.dispatchEvent(event);
    }
  }
```

Now, this method returns a render-event with a `WebGLRenderingContext` instead of a `CanvasRenderingContext2D`. So I extended the definition of a `RenderEvent` to allow for a webgl-context, too. But this now causes an issue in another place: `src/ol/render.js/getVectorContext` returns a `CanvasImmediateRenderer` that has a *2D* context. 

Obviously, I could update the type-annotations of `CanvasImmediateRenderer`, too, but this class is part of the `render/canvas` directory, which I think means that it is specific for the canvas-2d API. 

The consistent thing to do here, I believe, might be to create a `render/webgl` directory containing WebGL equivalents of all the functionality in `render/canvas`.  Is this the kind of structure that you'd like to see here? Or is there a simpler/better solution that I'm missing?

Thanks again!

**Comment by MichaelLangbein** [comment]
Allow me to rephrase the issue.
getVectorContext has a type-signature `Event -> CanvasImmediateRenderer`.

We could either:
 - create a separate `WebGlRenderEvent` that does not qualify as a input to `getVectorContext`
 - or create a class `WebGlCanvasImmediateRenderer`, that provides the same functionality as `CanvasImmediateRenderer`, but for a WebGL context. But I don't think that this is practical: creating functionality that creates (for example) a circle on the fly in WebGL would mean a lot of shader-meta-programming, which is not congruent with the idea of keeping WebGL-support lightweight.

**Comment by jahow** [comment]
I think we might as well not provide any context at all in the event. Having a webgl context isn't as useful as a canvas2d one, since the application code won't be able to do much with it anyway (without an immediate rendering API like in canvas2d).

Then using something like this should work for building the inverse transform: https://github.com/openlayers/openlayers/blob/f3a67e818289282ac71b6d13df96434dd44ace61/src/ol/renderer/canvas/TileLayer.js#L272-L289

Although I'm not sure that would be even useful considering there will be no drawing context provided anyway. @ahocevar what's your opinion on this?

**Comment by ahocevar** [comment]
If the purpose of this feature request is just to get notified on prerender and postrender, then I think we neither need a drawing context nor a drawing transform in the event. The existing event can be reused, the documentation for `context` is already correct, and the `inversePixelTransform` would get a similar documentation ("Only available when a Canvas renderer is used, null otherwise") and a `null` value when fired from a WebGL renderer.

**Comment by MichaelLangbein** [comment]
Roger that! Will adjust things accordingly. Thanks a lot!

**Comment by MichaelLangbein** [comment]
Hey guys!

I just noticed that my commit-message did not conform to the [PR-guidelines](https://github.com/openlayers/openlayers/blob/main/CONTRIBUTING.md#pull-request-guidelines). So I rebased my branch to catch up with the latest changes on master and chanded my commit-message, too. 

Now, however, I notice that netlify no longer builds correcty: 

```
1:21:09 PM: Build ready to start
1:21:11 PM: build-image version: b0258b965567defc4a2d7e2f2dec2e00c8f73ad6
1:21:11 PM: build-image tag: v3.4.1
1:21:11 PM: buildbot version: fcbc7a79f7372e5995cf02954ef19fc48d20c871
1:21:11 PM: Fetching cached dependencies
1:21:12 PM: Starting to download cache of 548.5MB
1:21:17 PM: Finished downloading cache in 5.505318973s
1:21:17 PM: Starting to extract cache
1:21:33 PM: Finished extracting cache in 16.440542025s
1:21:33 PM: Finished fetching cache in 22.082851237s
1:21:33 PM: Starting to prepare the repo for build
1:21:34 PM: Preparing Git Reference pull/11471/head
1:21:38 PM: Starting build script
1:21:38 PM: Installing dependencies
1:21:38 PM: Python version set to 2.7
1:21:38 PM: Started restoring cached node version
1:21:41 PM: Finished restoring cached node version
1:21:57 PM: Version '10' not found - try `nvm ls-remote` to browse available versions.
1:21:57 PM: Failed to install node version '10'
1:21:57 PM: Error running command: Build script returned non-zero exit code: 1
1:21:57 PM: Failing build: Failed to build site
1:21:57 PM: Failed during stage 'building site': Build script returned non-zero exit code: 1
1:21:57 PM: Finished processing build request in 46.088649215s
```

It seems that netlify attempts to restore an old node version that no longer exists. Could this be because my original PR is now 5 days old? What could I do to fix this? As always, thanks for all the help!

**Comment by MichaelLangbein** [comment]
Hi!
Alright, thanks a lot @ahocevar and @jahow  for all the help!
And thanks also for the [PR](https://github.com/openlayers/openlayers/pull/11471). Given that the [build issue](https://github.com/openlayers/openlayers/issues/11463#issuecomment-683412737) described above doesn't occur anymore, either, I have run out of things to complain about and guess we could close this issue :)

**Comment by tschaub** [comment]
Fixed by #11471.
